crates/extensions/tedge-p11-server/src/pkcs11.rs

```diff
@@ -78,38 +70,128 @@ impl Pkcs11SigningKey {
+            });
+        let (slot, _) = tokens
             .next()
             .context("Didn't find a slot to use. The device may be disconnected.")?;
-        let slot_info = pkcs11client.get_slot_info(slot)?;
-        let token_info = pkcs11client.get_token_info(slot)?;
+
+        let slot_info = self.context.get_slot_info(slot)?;
+        let token_info = self.context.get_token_info(slot)?;
         debug!(?slot_info, ?token_info, "Selected slot");
 
-        let session = pkcs11client.open_ro_session(slot)?;
+        let session = self.context.open_ro_session(slot)?;
         session.login(UserType::User, Some(pin))?;
         let session_info = session.get_session_info()?;
         debug!(?session_info, "Opened a readonly session");
 
-        let pkcs11 = PKCS11 {
-            session: Arc::new(Mutex::new(session)),
-        };
-
-        let key_type = get_key_type(pkcs11.clone()).context("Failed to read key")?;
-        trace!("Key Type: {:?}", key_type.to_string());
-        let key = match key_type {
-            KeyType::EC => Pkcs11SigningKey::Ecdsa(ECSigningKey { pkcs11 }),
-            KeyType::RSA => Pkcs11SigningKey::Rsa(RSASigningKey { pkcs11 }),
+        // get the signing key
+        let key = match &uri_attributes {
+            Some(uri::Pkcs11Uri { id, object, .. }) if id.is_some() || object.is_some() => {
+                let key = Self::find_key_by_attributes(&uri_attributes.unwrap(), &session)?;
+                let key_type = session
+                    .get_attributes(key, &[AttributeType::KeyType])?
+                    .into_iter()
+                    .next()
+                    .context("no keytype attribute")?;
+                let Attribute::KeyType(keytype) = key_type else {
+                    anyhow::bail!("can't get key type");
+                };
+                let pkcs11 = PKCS11 {
+                    session: Arc::new(Mutex::new(session)),
+                };
+                match keytype {
+                    KeyType::EC => Pkcs11SigningKey::Ecdsa(ECSigningKey { pkcs11 }),
+                    KeyType::RSA => Pkcs11SigningKey::Rsa(RSASigningKey { pkcs11 }),
+                    _ => panic!("unsupported key type"),
+                }
+            }
             _ => {
-                panic!("Unsupported key type. Only EC and RSA keys are supported");
+                let pkcs11 = PKCS11 {
+                    session: Arc::new(Mutex::new(session)),
+                };
+                let key_type = get_key_type(pkcs11.clone()).context("Failed to read key")?;
+                trace!("Key Type: {:?}", key_type.to_string());
+                match key_type {
+                    KeyType::EC => Pkcs11SigningKey::Ecdsa(ECSigningKey { pkcs11 }),
+                    KeyType::RSA => Pkcs11SigningKey::Rsa(RSASigningKey { pkcs11 }),
+                    _ => {
+                        panic!("Unsupported key type. Only EC and RSA keys are supported");
```

Comment:
Is a panic really appropriate here? I would return an error as this can be due to a misconfiguration.

Reply:
changed to `anyhow::bail!` in 0208319f